### PR TITLE
Nix build for a 'smart' docker image that can load keyfiles or a pier and boot a ship

### PR DESCRIPTION
**Pull request by eamsden**
This PR adds Nix and GitHub Actions builds for a docker image for Urbit, pushed to `tloncorp/urbit`

Builds are done alongside Urbit, Docker Hub pushes are done whenever a tag matching `urbit-v*` is pushed to GitHub/urbit/urbit

## Pre-merge requirements

- [x] I parameterized the name of the cachix cache, so `secrets.CACHIX_NAME` should be set to `ares`
- [x] `secrets.DOCKER_USERNAME` needs to be set to the docker hub username (`tloncorp` I think) for publishing, and `secrets.DOCKER_TOKEN` must be set to the *password* (not a PAT, otherwise README updating does not work, this is a docker hub limitation)

Once this is done the `build.yml` file will also build the docker image, and the `release-docker.yml` file will push the image to docker hub with a tag corresponding to the Urbit version, whenever a tag matching `urbit-v*` is pushed.

This PR tracks the progress on #1934 and my corresponding [proposal](https://grants.urbit.org/proposals/626138244-docker-image-for-cloud-container-orchestration)

## TODO list:
- [x] Nix: build for a docker image that can load keyfiles or a pier from a volume mount.
  * [x] Testing
    + [x] Boot from a .key file (tested on moons)
    + [x] Delete the container and make a new container, booting from the existing pier
    + [x] Touch a <name>.comet file and then make a container, resulting in a pier <name> for a new comet.
- [x] CI:
  * [x] Build docker image in CI (add to list of nix attributes built in `.github/workflows/build.yml`)
  * [x] Push docker image to a public Docker Hub repository (using [Skopeo](https://github.com/containers/skopeo) as per the Nixpkgs manual, add to `.github/workflows/release.yml`)
  * [x] Update docker hub readme with readme for our docker image (see [update-container-description-action](https://github.com/christian-korneck/update-container-description-action), add to `.github/workflows/release.yml`)
- [ ] Docs:
  * [x] README (to be visible on GitHub and Docker Hub, see CI push above, explaining how to boot from keyfiles and piers, and how to boot comets
    + [x] Ports to publish (HTTP/Eyre and UDP/Ames)
    + [x] Volume mountpoints and behavior (keyfile/pier/comet)
  * [ ] Azure Container Instances deployment guide - TBW on urbit.org
  * [ ] Amazon ECS deployment guide - TBW on urbit.org

### CODE REVIEW

**Comment by eamsden** [comment]
@brendanhay @jtobin can someone reach out to me about CI updates? I have github actions code sitting ready to go but I want to talk about how to stage/test it and whether this PR is the appropriate place. cc @jalehman

**Comment by jtobin** [comment]
@eamsden I think it should be fine to do that here.  Up to you!

**Comment by eamsden** [comment]
@jtobin since GitHub runs actions from `master` for external pulls, I'm going to test in my fork repo (not this PR) and post the results here after adding the workflow changes to the PR. I should have been more clear, my question was whether this PR was the appropriate place to include the workflow changes or whether I should PR them separately. Testing has to be done separately in any case for the aforementioned reason.

**Comment by jtobin** [comment]
All good.  I think it's fine to include the workflow changes here if you'd like.

**Comment by eamsden** [comment]
@jtobin this needs some updates to the secrets (detailed in an edit to the PR description above) for CI to run.

I'm going to make the Azure/AWS instructions hopefully tomorrow, and then I can take the draft tag off.

**Comment by jtobin** [comment]
@eamsden Done, I've set those for you.

**Comment by eamsden** [comment]
@jtobin I don't have permissions to kick the CI builds for the PR off again, can you do that?

**Comment by eamsden** [comment]
@jtobin all 4 builds are failing because the cachix action is missing the `name` input, but this is definitely specified in build.yml (and I built successfully in my own repo). Can you double-check the name and contents of the CACHIX_NAME secret?

**Comment by jtobin** [comment]
Right, I just checked the logs myself and noticed the same thing.  I've just checked and `CACHIX_NAME` is definitely set to `ares` -- I reassigned it, to be sure.

**Comment by eamsden** [comment]
I wonder if GitHub somehow limits access to secrets in PRs? That would be a nuisance. I will check the actions docs in a bit. You can have a look at my build.yml changes.

**Comment by jtobin** [comment]
> I wonder if GitHub somehow limits access to secrets in PRs? That would be a nuisance. I will check the actions docs in a bit. You can have a look at my build.yml changes.

Yeah, good question.  I can try pushing a branch of my own directly to the repo and see what happens?

**Comment by jtobin** [comment]
@eamsden Yeah, seems to be something like that.  Not sure if you need special perms to see these, but you can check the logs from the build of the branch copy [here](https://github.com/urbit/urbit/runs/1688437384?check_suite_focus=true).

**Comment by eamsden** [comment]
oddly it seems like the mac build now fails installing Nix, I wonder if it is because of adding the KVM option (necessary for building docker)? But that shouldn't mess up nix at install time.

**Comment by eamsden** [comment]
@jtobin I altered the build.yml file so it doesn't try to do extra config on mac. When you get a sec can you bump your test branch to that so we can try it with the secrets?

**Comment by jtobin** [comment]
:+1: bumped.

**Comment by eamsden** [comment]
Looks like those all worked. I will be ready to take the draft tag off later today then.

@jtobin @jalehman is it better for ECS/Container Instances docs to go here or as a PR to urbit.org docs about starting up a ship?

**Comment by eamsden** [comment]
@jtobin we should probably also do image signing but could do that as a separate PR, it will involve setting up 1 or 2 more secrets.

**Comment by eamsden** [comment]
@jtobin this is ready for review

**Review by jtobin — COMMENTED** [review]
The WIP commits should be fixed up and the messages reworded to match the repo's [general commit style](https://github.com/urbit/urbit/blob/master/CONTRIBUTING.md#hygiene), but I think this all looks pretty solid.  There's one comment re: things blowing up wanting kvm while building skopeo; is that still relevant, or did it only refer to the macOS build?

(I don't have anything to do with the related grant itself, so I'll tag @jalehman for a stamp with respect to that.)

**Comment by eamsden** [comment]
@jtobin  the need for the KVM feature tag is while building the docker image itself, not skopeo, but it does blow up the Mac build. That comment just explaining why I have two `using: cachix/install-nix-action` clauses with opposite `if` clauses. One runs install-nix-action on a linux build and adds the `kvm` tag to the nix config, which the docker image build checks for before running. If that tag isn't present then the docker image build will throw an error. This is because it doesn't actually use Docker but QEMU to build the docker image.

However, that line was causing the MacOS build to fail. We don't actually build the docker image on MacOS, as there is no point. Docker images for MacOS are still Linux docker images. So we don't need the `exta_nix_config` line. But the only way I could discover to do it conditionally was to make two, conditional, `uses` steps.

It's odd that it does blow up the MacOS nix install, as the config added is literally just a set of "tags" that nix looks in the config for a particular system for before building a particular derivation. But perhaps the install-nix-action will always blow up on MacOS with any `extra_nix_config`. Hard to say.

I hope that clarifies why the comment is there. I've also re-written the history to make it atomic and match the commit style as described.

**Comment by eamsden** [comment]
@jtobin one more push to get rid of the spurious comment discussed above.

**Review by jtobin — APPROVED** [review]

**Comment by eamsden** [comment]
@jtobin once this is approved, I wonder how we should go about getting the docker image released?

1. One option is to just wait for the next point-release of Urbit
2. Another is to backport these changes on top of urbit-v1.0 in a branch and then temporarily push a tag `urbit-v1.0docker` to trip the release-docker.yml build, and then expunge both the tag and branch.
3. The last option would be for someone with the docker hub credentials to manually run the description update command and skopeo. I can help with both if we pick this option

I personally favor option 2: it validates the CI work and ensures docker hub is immediately in sync with the latest release.

File: pkg/docker-image/README.md
Comment (jalehman) [inline pkg/docker-image/README.md:13]:
```suggestion
This image expects a volume mounted at `/urbit`. This volume should initially contain one of
```

File: pkg/docker-image/README.md
Comment (jalehman) [inline pkg/docker-image/README.md:17]: I'm having trouble understanding what exactly is meant by this sentence—could this be rephrased?

File: pkg/docker-image/README.md
Comment (jalehman) [inline pkg/docker-image/README.md:22]:
```suggestion
  * e.g. if your ship is `sampel-palnet` then you likely have a directory `sampel-palnet` whose path you pass to `./urbit` when starting. [Move your pier](https://urbit.org/using/operations/using-your-ship/#moving-your-pier) directory to the volume and then start the container.
```

**Review by jalehman — APPROVED** [review]
This looks great! Left a few small suggestions/comments on the README.

File: pkg/docker-image/README.md
Comment (eamsden) [reply pkg/docker-image/README.md]: I think this was a spurious paste as a result of writing docs too late at night. I've removed it.

**Comment by eamsden** [comment]
@jalehman I've integrated your suggestions and re-pushed to make sure the commit history is still properly atomic. cc @jtobin

**Comment by jtobin** [comment]
> Another is to backport these changes on top of urbit-v1.0 in a branch and then temporarily push a tag urbit-v1.0docker to trip the release-docker.yml build, and then expunge both the tag and branch.

This is fine by me.  In particular there have been no changes to Vere since the urbit-v1.0 release, so it should be fine to just push a tag to trigger the build (i.e., there is no need to backport anything).  I can do that post-merge.

**Comment by jtobin** [comment]
@eamsden I had to whitelist a few of the actions in the repository; will try to kick off release_docker again..

**Comment by jtobin** [comment]
[Success](https://hub.docker.com/r/tloncorp/urbit/tags?page=1&ordering=last_updated)!  Congrats!

**Comment by eamsden** [comment]
@jtobin thank you! I may open an issue or PR or two as I work through the docs and make screenshots, just to make sure everything is shipshape. One issue I see right now is that, instead of the behavior I expected, skopeo does *not* bump the "latest" tag by default. So I will have one quick follow-on PR for that soon.

**Pull request merged**
